game-core/src/main/java/games/strategy/triplea/ai/tree/BattleStep.java

```diff
@@ -0,0 +1,1221 @@
+      final List<Triple<Collection<Unit>, Collection<Unit>, Boolean>> groupsAndTargets,
+      final int groupsAndTargetsIndex,
+      final StepUnits currentUnits,
+      final Type type,
+      final MustFightBattle.ReturnFire returnFire,
+      final boolean allowMultipleHitsPerUnit,
+      final double probability) {
+    final List<Unit> firingGroup =
+        new ArrayList<>(groupsAndTargets.get(groupsAndTargetsIndex).getFirst());
+    final Collection<Unit> aliveUnits = currentUnits.getAliveFriendly();
+    final Collection<Unit> validAliveOrInjuredTargets =
+        groupsAndTargets.get(groupsAndTargetsIndex).getSecond();
+    final List<Unit> validAliveTargets =
+        groupsAndTargets.get(groupsAndTargetsIndex).getSecond().stream()
+            .filter(aliveUnits::contains)
+            .collect(Collectors.toList());
+
+    final boolean isSuicideOnHit = groupsAndTargets.get(groupsAndTargetsIndex).getThird();
+
+    final boolean defending = type == Type.DEFENDER || type == Type.AA_DEFENDER;
+    final RollData diceTuple;
+
+    if (type == Type.AA_ATTACKER || type == Type.AA_DEFENDER) {
+      diceTuple =
+          getAaRollData(
+              currentUnits,
+              defending,
+              allowMultipleHitsPerUnit,
+              firingGroup,
+              validAliveOrInjuredTargets);
+    } else {
+      diceTuple = getRegularRollData(defending, firingGroup, validAliveOrInjuredTargets);
+    }
+    final List<Double> hitProbabilities = calculateHitProbabilities(diceTuple);
+
+    int totalHits = 0;
+    double totalProbability = 0.0;
+    for (final Unit targetUnit : currentUnits.addFriendlyMultiHitTargets(validAliveTargets)) {
+
+      if (totalHits >= diceTuple.totalRolls) {
+        // no more rolls available
+        break;
+      }
+
+      final double hitProbability = hitProbabilities.get(totalHits);
+      totalProbability += hitProbability;
+      if (groupsAndTargetsIndex > 0) {
+        getTargetGroupFightOutcomes(
+            outcomes,
+            groupsAndTargets,
+            groupsAndTargetsIndex - 1,
+            new StepUnits(currentUnits),
+            type,
+            returnFire,
+            allowMultipleHitsPerUnit,
+            hitProbability * probability);
+      } else {
+        outcomes.compute(
+            new StepUnits(currentUnits),
+            (key, value) -> {
+              if (value == null) {
+                return hitProbability * probability;
+              } else {
+                return value + hitProbability * probability;
+              }
+            });
+      }
+
+      totalHits += 1;
+      if (MustFightBattle.ReturnFire.ALL == returnFire) {
+        currentUnits.hitFriendly(targetUnit);
+      } else {
+        currentUnits.killFriendly(targetUnit);
+      }
+      if (isSuicideOnHit) {
+        currentUnits.hitEnemy(firingGroup.get(totalHits - 1));
+      }
+    }
+
+    final double hitProbability = 1.0 - totalProbability;
+    if (groupsAndTargetsIndex > 0) {
```

Comment:
This logic seems very similar to line 496. Can you make a helper function with whatever params that differ?